src/dictionary.c

```diff
@@ -106,16 +146,41 @@ static int df_equal(const void* x, R_len_t i, const void* y, R_len_t j) {
   // `vec_proxy_equal()` flattens data frames so we don't need to
   // worry about df-cols
   for (R_len_t col = 0; col < n_col; ++col) {
-    if (!equal_scalar_p(types[col],
-                        R_NilValue, x_ptrs[col], i,
-                        R_NilValue, y_ptrs[col], j,
-                        true)) {
+    if (!equal_scalar_na_equal_p(types[col],
+                                 R_NilValue, x_ptrs[col], i,
+                                 R_NilValue, y_ptrs[col], j)) {
       return false;
     }
   }
 
   return true;
 }
+
+static int df_equal_missing(const void* x, R_len_t i) {
+  struct dictionary_df_data* x_data = (struct dictionary_df_data*) x;
+
+  enum vctrs_type* types = x_data->col_types;
+  void** x_ptrs = x_data->col_ptrs;
+  R_len_t n_col = x_data->n_col;
+
+  for (R_len_t col = 0; col < n_col; ++col) {
+    enum vctrs_type type = types[col];
+
+    // Raw doesn't have missing values
```

Comment:
Since raw vectors are probably pretty rare, and since there seems to be support for this in `equal_scalar_na_equal_p()`, could you let it just fall through to be simpler?

Reply:
The issues is that RAW vectors are not supported in `vec_type_missing_value()`.